proc_macros/src/lib.rs

```diff
@@ -44,3 +45,108 @@ pub fn builder_args(input: TokenStream) -> TokenStream {
     }
     .into()
 }
+
+struct InvalidRuleTestSpec {
+    code: Expr,
+    errors: ExprArray,
+}
+
+impl Parse for InvalidRuleTestSpec {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut code: Option<Expr> = Default::default();
+        let mut errors: Option<ExprArray> = Default::default();
+        let content;
+        braced!(content in input);
+        while !content.is_empty() {
+            let key: Ident = content.parse()?;
+            content.parse::<Token![=>]>()?;
+            match &*key.to_string() {
+                "code" => {
+                    code = Some(content.parse()?);
+                }
+                "errors" => {
+                    errors = Some(content.parse()?);
+                }
+                _ => panic!("didn't expect key {}", key),
+            }
+            if !content.is_empty() {
+                content.parse::<Token![,]>()?;
+            }
+        }
+        Ok(Self {
+            code: code.expect("Expected 'code'"),
+            errors: errors.expect("Expected 'errors'"),
+        })
+    }
+}
+
+impl ToTokens for InvalidRuleTestSpec {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let code = &self.code;
+        let errors = &self.errors;
+        quote! {
+            tree_sitter_lint::RuleTestInvalid::new(
```

Comment:
Like this does it seem weird that `tree_sitter_lint` is exposing these "rule-tester helper" types etc?

Reply:
I'm not sure. Is the main `tree_sitter_lint` crate something that anyone who's not writing a rule would actually use? Wouldn't they just run a binary? (I guess not, if custom rules have to be compiled.)

Reply:
Right this seems related to where I've now been heading/thinking about wrt "how do rules/plugins actually get "loaded"/run?"